### ISSUE TITLE
Please add an option to disable build workspace before rename Element

### ISSUE THREAD

**Issue by lidochca** [eclipse/xtext-eclipse#1188]
Hello, 

I had a discussion with Christian on xtext-dev mailing list about this topic
https://www.eclipse.org/lists/xtext-dev/msg01335.html 

We found that when we try to user 'Rename element' for our dsl files, it triggers workspace build. Our projects has a custom builder that executes a shell script that is time consuming.
It makes 'Rename Element' impossible to use.
We would need an option in preferences to disable/to avoid building the whole workspace when 
user tries to 'Rename Element'.
Just customizing our builder to avoid it's execution when nothing has changed, is not a solution for us.
We have project with mixed sources CPP/python/other custom languages, whose configuration paths and other files could change outside eclipse, by running an external builder, e.g. a ccmake from command line,  and when we trigger a build from eclipse, it still should happen event if for eclipse nothing has changed.
 
A simple option that would make possible to ovoid building the whole workspace would be great.
Something in SyncUtil.java like 

public void waitForBuild(IProgressMonitor monitor) {
try { 
      if ( workspace_build_enabled_during_rename) {
                workspace.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, monitor);
      }
} catch (CoreException e) {
      throw new OperationCanceledException(e.getMessage());
}
}
Could you please implement it in next xtext releases?

Thank you in advance.
Kind Regards
Lidia

**Comment by kthoms** [comment]
I'm strongly against a visible preference, since this requirement is extremely rare and unintended usage usually leads to an undesired behavior. 

However, preferences don't need to have a visible possibility for configuration in the workspace preferences. There are a couple of examples in Eclipse platform code where preferences are internally visible only. This might require some advanced knowledge of their presence and a custom provisioning of a workspace/project/bundle settings.

> Could you please implement it in next xtext releases?

I doubt that we will work on this soon, but we are open for contributions.

**Comment by lidochca** [comment]
Ok Thank you Karsten,
It is an acceptable solution for us. I can contribute with something like this:

IPreferenceStore prefs = Activator.getDefault().getPreferenceStore();
Boolean isDisabled = prefs.getBoolean(KEY_DISABLE_BUILD_DURING_RENAME);

And we can set the desired value in config.ini

Kind Regards
Lidia

**Comment by cdietrich** [comment]
i am very sceptic about that.
doing it as preference is not an option imho
maybe we can do a hidden secrect undocumented jvm option but i am very sceptic about that.

i also wonder if you have actually tried to "fix" the problem by extending and fixing the affected Xtext code

@szarnekow what do you think

**Comment by szarnekow** [comment]
I share the concerns of @cdietrich 

This will likely cause maintenance effort because questions in the direction of 'rename didn't do it' become more likely.

**Comment by lidochca** [comment]
The only change that I would need is in SynchUtil.waitForBuild()

```
public void waitForBuild(IProgressMonitor monitor) {
		try {
			IPreferenceStore prefs = Activator.getDefault().getPreferenceStore();
			Boolean isDisabled = prefs.getBoolean("SKIP_WAIT_FOR_BUILD");
			if (!isDisabled) {
				workspace.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, monitor);
			}
		} catch (CoreException e) {
			throw new OperationCanceledException(e.getMessage());
		}
	}
```

Can this be accepted?
We are using a plugin_customization.ini file and it will be sufficient for us to set this preference.

**Comment by cdietrich** [comment]
still dont have an answer to

> i also wonder if you have actually tried to "fix" the problem by extending and fixing the affected Xtext code

**Comment by lidochca** [comment]
@cdietrich I am not sure what you mean. 
You mean to extend the SyncUtil class in my project, to implement my own version or Rename Refactor ?

**Comment by cdietrich** [comment]
Subclass and Bind the classes that call the util and don’t call the util

**Comment by lidochca** [comment]
Thanks @cdietrich .
I tried to implement what you suggested about subclass and bind, but I am not very familiar how injection is used in xtext, and did not found yet how to override Rename Element action inherited from xtext. I can still give a try, but it already seems that many classes should be extended and maintained, at least those functions that should be overridden.
I think that better and easier solution for us would still be to be able to customize to skip the build before rename like I just committed.

**Comment by kthoms** [comment]
@lidochca Your patch shows that the desired behavior can be achieved by a customizing of class `SyncUtil`. Since instances of that class are injected in Xtext code, you can provide a customized implementation. So just subclass `SyncUtil`, create a Guice Module that binds `SyncUtil.class` to `YourSyncUtil.class` and register this Guice module with the `org.eclipse.xtext.ui.shared.overridingGuiceModule` extension point.
The use case you describe here is a corner case and IMHO should not be encouraged by direct framework support.
Please try to make Xtext use your customized class in your specific project context. If that works, please close the PR.

**Comment by lidochca** [comment]
Thanks @kthoms 
I will try your suggestions and will let you know.
Sounds as a good idea for me, if it will work.
Regards
Lidia

**Comment by lidochca** [comment]
Hello Karsten @kthoms 
I am struggling to make it work based on your last recommendations.
I tried few different combinations, nothing helps to load function waitForBuild() from my version of SyncUtil class. Could your please help me?
I did
` <extension
         point="org.eclipse.xtext.ui.shared.overridingGuiceModule">
      <module
            class="org.eclipse.xtext.ui.rename.MyModule">
      </module>
   </extension>`

`public class MyModule extends AbstractGenericModule {
	@Override
	public void configure(Binder binder) {
		System.out.println("My configure "+ binder);
		binder.bind(SyncUtil.class).to(MySyncUtil.class);
	}
}`

`public class MySyncUtil extends SyncUtil {

	@Inject(optional = true)
	private IWorkspace workspace;
		
	private static final boolean SKIP_WAIT_FOR_BUILD = Boolean.getBoolean("my.xtext.skipWaitForBuild");

  @Override
	public void waitForBuild(IProgressMonitor monitor) {
		try {
			System.out.println("My waitForBuild ");
			if ( !SKIP_WAIT_FOR_BUILD) {
				workspace.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, monitor);
			}
		} catch (CoreException e) {
			throw new OperationCanceledException(e.getMessage());
		}
	}
}`

I can see the log from MyModule.configure(), but not the log from MySyncUtil.waitForBuild()

Thank you
Kind Regards
Lidia

**Comment by cdietrich** [comment]
which is the place/syncutil call you try to fix. are you sure its not lanuage specific and created via mydsluimodule=

**Comment by lidochca** [comment]
I posted my changes on

https://github.com/lidochca/xtext.skip.build.before.rename
https://github.com/lidochca/xtext.skip.build.before.rename/blob/master/plugin.xml
https://github.com/lidochca/xtext.skip.build.before.rename/blob/f2da1af5f71b8cc1da9216b49cfd1206bb3fc367/src/org/eclipse/xtext/ui/rename/MyModule.java
https://github.com/lidochca/xtext.skip.build.before.rename/blob/f2da1af5f71b8cc1da9216b49cfd1206bb3fc367/src/org/eclipse/xtext/ui/rename/MySyncUtil.java

**Comment by cdietrich** [comment]
yes but why in overridingGuiceModule and not in MyDslUiModule if the rename participant is created by that

**Comment by lidochca** [comment]
I am trying to override SyncUtil.waitForBuild();
I think was not created thought mydsluimodule= , at least I did not found such references.

**Comment by cdietrich** [comment]
what about simply adding a binding and give it a try?

**Comment by kthoms** [comment]
I could reproduce your issue with your example, thanks. A problem here might be that you are trying to influence Xtext's behavior (rename refactoring) itself, not your own language.
I'm investigating this now a bit.

**Comment by lidochca** [comment]
Thanks a lot for your help, I also will see what else I can do, and let you know how it goes.

**Comment by cdietrich** [comment]
Closed as wont-fix. Feel free to reopen at github.com/eclipse/xtext
